### PR TITLE
Topological Sorting and Cycle Detection for Workflow (WIP)

### PR DESCRIPTION
**Pull request by geetduggal**
@orodeh Thanks for the intro this afternoon. This is a WIP PR as I develop this pass.  

The goal of the PR is to perform a sorting of the calls, scatters, and declarations in the WDL workflow as a preprocessing step to avoid compiler failure on forward references.   A side-benefit is returning an error on cycle detection in the directed graph.

The current state reorders the calls correctly on a couple of sample files, but there's still a bit more work to be done to ensure any scatter or declaration is handled properly not to mention adding tests.   I also would like to see/test if I can use the `GraphNode` class to handle all Scope objects of relevance in the topological sort. Note that currently I don't expect builds to pass as I am simply exiting the compilation after the preprocessing step. 

@mlin interest.

### CODE REVIEW

**Comment by orodeh** [comment]
@geetduggal, glad you were able to start coding so quickly!

When you feel it is ready, please let me know, and I will take a through look at it.

**Comment by geetduggal** [comment]
The topological sorting now has functionality for what I call 'scatter collapsing'.  This in effect rewires the graph so that (1) any descendant of a scatter is treated as the scatter itself and (2) any other node pointing to a descendant node in the scatter points to the scatter itself.

This commit then allows for more intuitive sorting of the graph so that the order of operations can be roughly understood by the order of the top level calls and scatters within the file.  It has the added benefit of performing directed cycle detection at the top level. For example, if a workflow contains a scatter S that is dependent on some call A and call A is dependent on some different call B within the scatter S, it will treat this as a cycle.   While this is not strictly a cycle in terms of calls to tasks in the DAG, this is desirable particularly from a readability perspective because it becomes more difficult to understand the sequence of the workflow that could have many such upstream dependencies.

In other words, for execution on the platform, we require that a DAG be built from the top level components in a WDL workflow.

**Comment by geetduggal** [comment]
So the current state correctly performs the recursive sorting and in a way that conceptually makes sense.  This being said, the method is unfortunately still a bit complex (this complexity was only arrived after finding flaws in simpler attempts). I'm sure that in some other more idealized setting, the framework could be simplified a bit.  After the memorial day holiday I'll talk to @orodeh and @mlin about potential simplifications and optimizations.

**Comment by orodeh** [comment]
Geet, I've given you permissions to the original dxWDL git project. You should be able to work with it directly. 

I am not sure I understand why the travis build fails, but we can discuss it.

**Comment by geetduggal** [comment]
Updated the code to perform an alternative sort as well as the draft documentation with info re: this. Adding @commandlinegirl to the PR as she expressed interest may take on a subsequent dxWDL related task to learn how it works.

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  *
+  *  Instead of handling expressions in calls directly,
+  *  we lift the expressions, generate auxiliary variables, and
+  *  call the task with values or variables (no expressions).
+  *
+  *  A difficulty we face here, is avoiding using internal
+  *  representations used by wdl4s. For example, we want to reorganize
+  *  scatter blocks, however, we cannot create valid new wdl4s scatter
+  *  blocks. Instead, we pretty print a new workflow and then load it.
+  */
```
Comment (orodeh) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: This comment is outdated.

**Comment by orodeh** [comment]
@geetduggal, I am going through this pull request, sorry for the extended delay. 

**TopoSort.md**
It is excellent to have an overall description of what this module does, and what it is used for. I printed and marked some editorial suggestions (depending on taste).

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            // Partition the set of nodes into those that have a predecessor and those that do not
+            // (Actually this will partition the map into two)
+            val (noPreds, hasPreds) = toPreds.partition { keyval =>
+                val predSet = keyval._2
+                predSet.isEmpty
+            }
+
+            // If there are no nodes left with no predecessors, either we are done recursing or there is a cycle
+            if (noPreds.isEmpty) {
+
```
Comment (orodeh) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Personally, I prefer to minimize the number of empty lines. The issue is that methods become very long, violating the rule that they should fit into one editor page.

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                // List of nodes with no predecessors
+                val noPredNodes = noPreds.map { keyval => keyval._1 }
+
+                // Build a new list of predecessors that do not contain the nodes above
+                val newToPreds = hasPreds.mapValues { predSet => predSet -- noPredNodes }
+                tsort(newToPreds, done ++ noPredNodes)
+            }
+        }
+
```
Comment (orodeh) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Your changes make this code **much** more readable, thanks!

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            acc + (e._1 -> acc.getOrElse(e._1, Set())) + (e._2 -> (acc.getOrElse(e._2, Set()) + e._1))
+        }
+        tsort(toPred, Seq())
+    }
+
+    // Internal procedure to sort children nodes at a particular level in the AST
+    def tsortASTnodes(nodes: Seq[Scope], cState:State, recursionDepth: Int) : Seq[Scope] = {
+        val indent = " "*recursionDepth*4
+
+        val recursedNodes = nodes.map {
```
Comment (orodeh) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Can you specify the type of recurseNodes?

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        // e.g. suppose call x has dependencies a,b,c.  This procedure will generate:
+        // [(a,x), (b,x), (c,x)]
+        val edges : Set[(Scope, Scope)] = recursedNodes.map { gnode =>
+            val nodeDependencies : Vector[(Scope, Scope)] = (gnode.upstream).map { dependency =>
+               val dependencyActual : Scope =
+                   // If any node's parent is a descendant of a root
+                   // level scatter, use the scatter as the parent
+                   if (scatterParent.contains(dependency)) {
+                       scatterParent(dependency).asInstanceOf[Scope]
+                   }
```
Comment (orodeh) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: The if/else style used, currently, is: 
if (cond) {
} else {
}

It would be preferable, to have all the code use the same style.

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+                    ((upstreamParents -- nodeDescendants) - gnode).map{ parent =>
+                        // If the parent outside the scatter is also a member of a scatter, use it instead.
+                        if ( scatterParent.contains(parent) && gnode != scatterParent(parent)) {
+                            (scatterParent(parent), gnode)
+                        } else {
+                            (parent, gnode)
+                        }
+                    }.toVector
+                }
+                case anyOtherNode => Vector[(Scope, Scope)]()
```
Comment (orodeh) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Elsewhere, I use the syntax:
case _ => Vector[(Scope, Scope)]()

I am fine with this style, though.

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                   else {
+                       dependency.asInstanceOf[Scope]
+                   }
+               (dependencyActual, gnode.asInstanceOf[Scope])
+            }.toVector
+
+            val descendantParents : Vector[(Scope, Scope)] = gnode match {
+                case scatter: Scatter => {
+                    // If a scatter's descendants have parents outside the scatter context,
+                    // map the scatter itself to them.
```
Comment (orodeh) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: You have an example for when this code is needed in TopoSort.md. Can you point to it from here?

**Comment by orodeh** [comment]
@geetduggal This is good stuff, and the issues are minor. After fixing, we should dit down and merge the code.

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala (hunk 1 above)
Comment (geetduggal) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Removed. Thanks!

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala (hunk 2 above)
Comment (geetduggal) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Thanks -- minimizing :)

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala (hunk 3 above)
Comment (geetduggal) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: thanks!

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala (hunk 4 above)
Comment (geetduggal) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Done.  I think it helps make it more readable. thx

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala (hunk 5 above)
Comment (geetduggal) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Done.

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala (hunk 7 above)
Comment (geetduggal) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Done.

File: src/main/scala/dxWDL/CompilerTopologicalSort.scala (hunk 6 above)
Comment (geetduggal) [inline src/main/scala/dxWDL/CompilerTopologicalSort.scala]: Done.

**Comment by orodeh** [comment]
LGTM!

**Comment by orodeh** [comment]
Merged!
